scripts/install-nix-from-closure.sh

```diff
@@ -1,149 +1,635 @@
+    # Form:
+    ### Application: Prefix: Body of message
+    printf '%s%s: %s: %s%s\n' "$color" "$appname" "$prefix" "$message" "$reset"
+
+    # Output is done, unset variables that could've been set by parent functions
+    unset message
+    unset color
+    unset prefix
+}
+
+notice() {
+    message="$1"
+    color="$green"
+    prefix='Notice'
+    print
+}
+
+warning() {
+    message="$1"
+    color="$yellow"
+    prefix='Warning'
+    >&2 print
+}
+
+error() {
+    message="$1"
+    exitSig="$2"
+    color="$red"
+    prefix='Error'
+    >&2 print
+    if [ -z "$exitSig" ]; then
+        exit 1
+    fi
+}
+
+errorRevert() {
+    message="$1"
+    color="$red"
+    prefix='Error'
+    >&2 print
+}
+
+contactme() {
+    print '
+
+    If you hit a bug, we would love to help.
+
+    You can, search/open an issue at https://github.com/nixos/nix/issues
+
+    Feel free to contact the team,
+     - on IRC: #nixos on irc.freenode.net
+     - on twitter: @nixos_org
+     - on Matrix: https://riot.im/app/#/room/#nix:matrix.org
+'
+}
+
+}
+###############################
+###  Program stage functions
+###############################
+{
+checkingRequirements() {    # This function must have checks only. Without making ANY changes to the system.
+    if ! [ -e "$self/.reginfo" ]; then
+        error "
+
+    Installer is incomplete ($self/.reginfo is missing)
+    "
+    fi
 
-set -e
+    if [ "$(id -u)" -eq 0 ]; then
+        notice "
 
-dest="/nix"
-self="$(dirname "$0")"
-nix="@nix@"
-cacert="@cacert@"
+    Install executed for ROOT.
 
+    Doing classic Linux package manager mode.
+    In Nix this mode is called: single-user mode for root.
 
-if ! [ -e "$self/.reginfo" ]; then
-    echo "$0: incomplete installer (.reginfo is missing)" >&2
-fi
+    Nix can do multi-user mode, and manage package trees for users independently.
+    "
+    fi
 
-if [ -z "$USER" ]; then
-    echo "$0: \$USER is not set" >&2
-    exit 1
-fi
+    # In case if USER not set
+    if [ -z "$USER" ]; then
+        notice "
 
-if [ -z "$HOME" ]; then
-    echo "$0: \$HOME is not set" >&2
-    exit 1
-fi
+    Environment variable USER is not set.
+    Maybe running inside container.
+    "
+        USER="$(id -u -n)"    # id is POSIX
+        print "Detected username: $USER"
+    fi
 
-# macOS support for 10.10 or higher
-if [ "$(uname -s)" = "Darwin" ]; then
-    if [ $(($(sw_vers -productVersion | cut -d '.' -f 2))) -lt 10 ]; then
-        echo "$0: macOS $(sw_vers -productVersion) is not supported, upgrade to 10.10 or higher"
-        exit 1
+    if [ -z "$HOME" ]; then
+        error "
+
+    Environment variable HOME is not set.
+    "
     fi
 
-    printf '\e[1;31mSwitching to the Multi-User Darwin Installer\e[0m\n'
-    exec "$self/install-darwin-multi-user"
-    exit 0
-fi
+    # macOS support for 10.10 or higher
+    if [ "$(uname -s)" = 'Darwin' ]; then
+        if [[ $(($(sw_vers -productVersion | cut -d '.' -f 2))) -lt 10 ]]; then
```

Comment:
`[[ ]]` is a bashism

Reply:
This is already part of Graham MacOS code.
MacOS by default always uses Bash. And Graham's script fully uses Bash.

I not changed this because it does not have sense to change to POSIX.

Reply:
I just can't check what:
```bash
if [[ $(($(sw_vers -productVersion | cut -d '.' -f 2))) -lt 10 ]]
```
Parses and does.

I do not understand this Bash:
```bash
if [[ $(($(val))) -lt 10 ]]]; then
```

What is the sense of:
```bash
$(($(val)))
```
I don't know what is $(()). Is it calculation and taking value?
I don't have macOS to see input.

Reply:
`$(())` is for arithmetic. Eg:
```
$ x=$((4 + 5));
$ echo "$x"
9
```

In that case I think that it can be ignored since `$(sw_vers -productVersion | cut -d '.' -f 2)` would return a single number. The only difference would be if it didn't return a number. Eg:
```
$ x=$(($(echo foo)))
$ echo $x
0
```

Reply:
That is why using logic I thrown my hands.

`$(())` must calculate and return number, but I don't know what policy and history Apple has with `sw_vers -productVersion` on macOS, and why and how it can happen that it calculates some numbers inside it.